anndata/core/anndata.py

```diff
@@ -1416,16 +1416,22 @@ def _remove_unused_categories(self, df_full, df_sub, uns):
             all_categories = df_full[k].cat.categories
             df_sub[k].cat.remove_unused_categories(inplace=True)
             # also correct the colors...
-            if f'{k}_colors' not in uns:
+            color_key = f"{k}_colors"
+            if color_key not in uns:
                 continue
+            color_vec = uns[color_key]
             # this is a strange hack...
-            if np.array(uns[f'{k}_colors']).ndim == 0:
-                idx = None
+            # What case is this? Does this happen?
+            if np.array(color_vec).ndim == 0:
+                uns[color_key] = np.array(color_vec)[(None,)]
+            elif len(color_vec) < len(all_categories):
```

Comment:
Maybe `elif len(color_vec) != len(all_categories):` instead?

Reply:
The current behavior is to re-use the colors in that case, but I've never been sure if that's intentional. It preserves user supplied color schemes, but it doesn't necessarily preserve the mapping between cluster identifier and color.

Reply:
You two are right: It doesn’t makes sense that way. They should just be redone.